nixos/modules/tasks/network-interfaces.nix

```diff
@@ -376,20 +376,10 @@ in
 
     networking.hostName = mkOption {
       default = "nixos";
-      # Only allow hostnames without the domain name part (i.e. no FQDNs, see
-      # e.g. "man 5 hostname") and require valid DNS labels (recommended
-      # syntax). Note: We also allow underscores for compatibility/legacy
-      # reasons (as undocumented feature):
-      type = types.strMatching
-        "^$|^[[:alpha:]]([[:alnum:]_-]{0,61}[[:alnum:]])?$";
+      type = types.str;
       description = ''
-        The name of the machine. Leave it empty if you want to obtain it from a
-        DHCP server (if using DHCP). The hostname must be a valid DNS label (see
-        RFC 1035 section 2.3.1: "Preferred name syntax") and as such must not
-        contain the domain part. This means that the hostname must start with a
-        letter, end with a letter or digit, and have as interior characters only
-        letters, digits, and hyphen. The maximum length is 63 characters.
-        Additionally it is recommended to only use lower-case characters.
```

Comment:
I would still keep that part or a similar version as a recommendation (or at least refer to `hostname(5)`).